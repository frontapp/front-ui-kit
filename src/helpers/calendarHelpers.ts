import {range} from 'lodash';
import {DateTime} from 'luxon';

/*
 * Constants.
 */

export enum CalendarWeekDaysEnum {
  SUNDAY = 'sunday',
  MONDAY = 'monday',
  TUESDAY = 'tuesday',
  WEDNESDAY = 'wednesday',
  THURSDAY = 'thursday',
  FRIDAY = 'friday',
  SATURDAY = 'saturday'
}

export const calendarWeekdayToWeekdayNumber = {
  [CalendarWeekDaysEnum.MONDAY]: 1,
  [CalendarWeekDaysEnum.TUESDAY]: 2,
  [CalendarWeekDaysEnum.WEDNESDAY]: 3,
  [CalendarWeekDaysEnum.THURSDAY]: 4,
  [CalendarWeekDaysEnum.FRIDAY]: 5,
  [CalendarWeekDaysEnum.SATURDAY]: 6,
  [CalendarWeekDaysEnum.SUNDAY]: 7
};

/*
 * Helpers.
 */

/** Get the sorted weekdays number, starting with Sunday. */
export function getSortedWeekdays(calendarWeekStartDay: CalendarWeekDaysEnum) {
  const sortedCalendarWeekdays = getSortedWeekdayNames(calendarWeekStartDay);
  return sortedCalendarWeekdays.map(weekdayName => calendarWeekdayToWeekdayNumber[weekdayName]);
}

export function getSortedWeekdayNames(calendarWeekStartDay: CalendarWeekDaysEnum) {
  const unorderedWeekdays = Object.values(CalendarWeekDaysEnum);
  const splitIndex = unorderedWeekdays.indexOf(calendarWeekStartDay);
  const left = unorderedWeekdays.slice(0, splitIndex);
  const right = unorderedWeekdays.slice(splitIndex);
  return [...right, ...left];
}

/** Get the full list of days for the provided month, sorted by the provided week days. */
export function getCalendarDays(month: DateTime, weekdays: ReadonlyArray<number>) {
  // Find the first datetime of the months we're going to work with.
  const previousMonth = month.minus({months: 1});
  const nextMonth = month.plus({months: 1});

  // Figure out the size of the left & right padding.
  const startOfmonth = month.startOf('month');
  const previousPad = weekdays.indexOf(startOfmonth.weekday);
  const rows = Math.ceil((month.daysInMonth + previousPad) / 7);
  const totalBoxes = rows * 7;
  const nextPad = totalBoxes - (previousPad + month.daysInMonth);

  // Build the day numbers for each range.
  const previousMonthDayNumbers = range(
    previousMonth.daysInMonth - previousPad + 1,
    previousMonth.daysInMonth + 1
  );
  const currentMonthDayNumbers = range(1, month.daysInMonth + 1);
  const nextMonthDayNumbers = range(1, nextPad + 1);

  // Map the day numbers to datetimes.
  const previousMonthDays = previousMonthDayNumbers.map((day: number) => previousMonth.set({day}));
  const currentMonthDays = currentMonthDayNumbers.map((day: number) => month.set({day}));
  const nextMonthDays = nextMonthDayNumbers.map((day: number) => nextMonth.set({day}));

  // Combine everything and return.
  return [...previousMonthDays, ...currentMonthDays, ...nextMonthDays];
}
