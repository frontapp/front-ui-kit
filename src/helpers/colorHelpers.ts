/*
 * Interfaces.
 */

interface FullColorPalette {
  readonly white: string;
  readonly shade10: string;
  readonly shade20: string;
  readonly shade30: string;
  readonly shade40: string;
  readonly shade50: string;
  readonly shade60: string;
  readonly shade70: string;
  readonly shade80: string;
  readonly shade90: string;
  readonly black: string;
}

export interface PaletteColor {
  readonly shade10: string;
  readonly shade20: string;
  readonly shade30: string;
  readonly shade40: string;
  readonly shade50: string;
}

type Palette = {readonly [T in PaletteColorsEnum]: PaletteColor};

/*
 * Constants.
 */

export enum PaletteColorsEnum {
  GREY = 'grey',
  PINK = 'pink',
  RED = 'red',
  ORANGE = 'orange',
  YELLOW = 'yellow',
  GREEN = 'green',
  TEAL = 'teal',
  BLUE = 'blue',
  PURPLE = 'purple'
}

export const greys: Readonly<FullColorPalette> = {
  white: '#ffffff',
  shade10: '#f7f8fa',
  shade20: '#f3f4f6',
  shade30: '#e8eaed',
  shade40: '#d4d6d9',
  shade50: '#bbbdbf',
  shade60: '#9b9c9e',
  shade70: '#77787a',
  shade80: '#3e3e40',
  shade90: '#19191a',
  black: '#001B38'
};

export const palette: Palette = {
  [PaletteColorsEnum.GREY]: {
    shade10: greys.shade10,
    shade20: greys.shade30,
    shade30: greys.shade50,
    shade40: greys.shade70,
    shade50: greys.shade80
  },
  [PaletteColorsEnum.PINK]: {
    shade10: '#fdf0f8',
    shade20: '#fbe1f2',
    shade30: '#f8cae8',
    shade40: '#ca3e99',
    shade50: '#9f3079'
  },
  [PaletteColorsEnum.RED]: {
    shade10: '#fff1f1',
    shade20: '#ffe1de',
    shade30: '#fac1bb',
    shade40: '#e9483a',
    shade50: '#b72215'
  },
  [PaletteColorsEnum.ORANGE]: {
    shade10: '#fff7f0',
    shade20: '#ffe9d1',
    shade30: '#fadab9',
    shade40: '#f2830b',
    shade50: '#cc6000'
  },
  [PaletteColorsEnum.YELLOW]: {
    shade10: '#fefbef',
    shade20: '#fef6dc',
    shade30: '#faebbb',
    shade40: '#e3b51e',
    shade50: '#91691b'
  },
  [PaletteColorsEnum.GREEN]: {
    shade10: '#eefbf3',
    shade20: '#d9f6e3',
    shade30: '#b6f0ca',
    shade40: '#10aa40',
    shade50: '#0c8331'
  },
  [PaletteColorsEnum.TEAL]: {
    shade10: '#f0fdff',
    shade20: '#dcfbff',
    shade30: '#b9f2fa',
    shade40: '#15acc0',
    shade50: '#0d7482'
  },
  [PaletteColorsEnum.BLUE]: {
    shade10: '#f1f6fd',
    shade20: '#e0edff',
    shade30: '#bed7fa',
    shade40: '#367fee',
    shade50: '#2356b3'
  },
  [PaletteColorsEnum.PURPLE]: {
    shade10: '#f8f1ff',
    shade20: '#f0dfff',
    shade30: '#ddbbfa',
    shade40: '#9235e4',
    shade50: '#681aad'
  }
};
