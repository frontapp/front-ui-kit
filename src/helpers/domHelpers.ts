export function getWidthWithMarginOfElement(node: Element) {
  if (!window || !window.document) return 0;

  const styles = window.getComputedStyle(node);
  const width = getPropertyAsNumber(styles, 'width') || 0;
  const marginLeft = getPropertyAsNumber(styles, 'margin-left') || 0;
  const marginRight = getPropertyAsNumber(styles, 'margin-right') || 0;

  return width + marginLeft + marginRight;
}

function getPropertyAsNumber(computedStyle: CSSStyleDeclaration, propertyName: string): number | null {
  const valueAsString: string = computedStyle.getPropertyValue(propertyName);
  const valueMatch = valueAsString.match(/(.*)px$/); // ComputedStyle calculates everything in pixels. e.g. '12.3px'.
  if (!valueMatch) return null;

  const value = Number(valueMatch[1]);
  return isNaN(value) ? null : value;
}
