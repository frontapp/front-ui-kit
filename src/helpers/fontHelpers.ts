/*
 * Interfaces.
 */

interface FontSizes {
  readonly veryTiny: string;
  readonly tiny: string;
  readonly verySmall: string;
  readonly small: string;
  readonly medium: string;
  readonly large: string;
  readonly veryLarge: string;
}

interface FontWeights {
  readonly thin: number;
  readonly light: number;
  readonly normal: number;
  readonly medium: number;
  readonly semibold: number;
  readonly bold: number;
  readonly black: number;
}

interface Fonts {
  readonly system: string;
  readonly monospace: string;
}

/*
 * Constants.
 */

export const fontSizes: FontSizes = {
  veryTiny: '10px',
  tiny: '11px',
  verySmall: '12px',
  small: '13px',
  medium: '14px',
  large: '16px',
  veryLarge: '18px'
};

export const fontWeights: FontWeights = {
  thin: 100,
  light: 300,
  normal: 400,
  medium: 500,
  semibold: 600,
  bold: 700,
  black: 800
};

export enum VisualSizesEnum {
  SMALL = 'small',
  MEDIUM = 'medium',
  LARGE = 'large',
  EXTRA_LARGE = 'extra-large'
}

export const fonts: Fonts = {
  system: `
    -apple-system,
    BlinkMacSystemFont,
    "Segoe UI",
    "Helvetica Neue",
    Helvetica,
    Arial,
    sans-serif,
    "Apple Color Emoji",
    "Segoe UI Emoji",
    "Segoe UI Symbol"
  `,
  monospace: `
    monaco,
    Consolas,
    "Lucida Console",
    monospace
  `
};
