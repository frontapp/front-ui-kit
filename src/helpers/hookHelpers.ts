import {useCallback, useEffect, useRef} from "react";

/*
 * Use Timeout.
 */

/** Returns a setTimeout-like function that clears the previous timeout when re-invoked or when the component unmounts. */
export function useTimeout(): [(handler: () => void, timeout?: number) => void, () => void] {
  const handleRef = useRef<number>();

  useEffect(() => () => clearTimeout(handleRef.current), []);

  const safeSetTimeout = useCallback((handler: () => void, timeout?: number) => {
    clearTimeout(handleRef.current);
    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions, @typescript-eslint/no-explicit-any
    handleRef.current = setTimeout(handler, timeout) as any;
  }, []);
  const safeClearTimeout = useCallback(() => clearTimeout(handleRef.current), []);

  return [safeSetTimeout, safeClearTimeout];
}

/** Returns the passed value as it was after the last successful render of the component. */
export function usePrevious<T>(value: T) {
  const ref = useRef<T | null>(null);
  useEffect(() => {
    ref.current = value;
  });
  return ref.current;
}
