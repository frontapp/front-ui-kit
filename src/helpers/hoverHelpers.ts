/*
 * Constants.
 */

const hoverParentClass = 'front-hover-parent';
const hoverClass = 'front-hover';

export const hoverSelector = buildHoverSelector();
export const directHoverSelector = buildDirectHoverSelector();

/*
 * Methods.
 */

export function buildHoverSelector(pseudoElement = '') {
  return `
    .${hoverParentClass}:hover &&${pseudoElement},
    .${hoverParentClass}.${hoverClass} &&${pseudoElement}
  `;
}

export function buildDirectHoverSelector(pseudoElement = '') {
  return `
    &.${hoverParentClass}:hover${pseudoElement},
    &.${hoverParentClass}.${hoverClass}${pseudoElement},
    .${hoverParentClass}:hover > &&${pseudoElement},
    .${hoverParentClass}.${hoverClass} > &&${pseudoElement}
  `;
}

export function buildHoverParentClassName(hasHover?: boolean, baseClassName: string = '') {
  const className = `${hoverParentClass}${baseClassName && ` ${baseClassName}`}`;
  return hasHover ? `${hoverClass} ${className}` : className;
}
