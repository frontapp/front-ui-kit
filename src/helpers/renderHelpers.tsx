import React from "react";
import * as ReactIs from "react-is";

/*
 * Helpers.
 */

/**
 * Takes in children for a component and looks for a specified component display name.
 */
export function isComponentInChildren(children: React.ReactNode, componentDisplayName: string) {
  return React.Children.toArray(children).some(child => {
    if (typeof child === 'string' || typeof child === 'number')
      return null;
    if (ReactIs.isFragment(child) || !ReactIs.isElement(child))
      return false;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/consistent-type-assertions
    return (child.type as any)?.displayName === componentDisplayName;
  });
}

/**
 * Takes in children for a component and strips out all content besides the first Icon.
 */
export function renderFirstIconOnly(children: React.ReactNode, shouldDisableColor: boolean = true) {
  let hasFoundIcon = false;
  return React.Children.toArray(children).map(child => {
    if (typeof child === 'string' || typeof child === 'number')
      return null;
    if (ReactIs.isFragment(child) || !ReactIs.isElement(child))
      return null;

    // Check for an icon and if we find one make sure we only have 1 icon
    // total that we could render. We also need to display any set colors
    // so the component itself can style the color.
    // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/consistent-type-assertions
    if ((child.type as any)?.displayName === 'Icon' && !hasFoundIcon) {
      hasFoundIcon = true;
      return React.cloneElement(child, {shouldDisableColor});
    }
    return null;
  });
}

/**
 * Takes in children for a component and strips out any specified component.
 */
export function renderChildrenIgnoreSpecifiedComponents(children: React.ReactNode, componentsToIgnore: ReadonlyArray<string>) {
  return React.Children.toArray(children).map(child => {
    if (typeof child === 'string' || typeof child === 'number')
      return child;
    if (ReactIs.isFragment(child) || !ReactIs.isElement(child))
      return child;

    // Check if the display name is one we should not render.
    // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/consistent-type-assertions
    if (componentsToIgnore.includes((child.type as any)?.displayName))
      return null;

    return child;
  });
}

/**
 * Takes in children for a component and renders only the specified components.
 */
export function renderChildrenSpecifiedComponents(children: React.ReactNode, componentsToInclude: ReadonlyArray<string>) {
  return React.Children.toArray(children).map(child => {
    if (typeof child === 'string' || typeof child === 'number')
      return null;
    if (ReactIs.isFragment(child) || !ReactIs.isElement(child))
      return null;

    // Check if the display name is one we should not render.
    // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/consistent-type-assertions
    if (componentsToInclude.includes((child.type as any)?.displayName))
      return child;

    return null;
  });
}
