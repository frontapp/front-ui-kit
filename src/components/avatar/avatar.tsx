import React, {FC, useMemo} from 'react';
import styled, {css} from 'styled-components';

import {greys, palette, PaletteColorsEnum} from '../../helpers/colorHelpers';
import {fonts, fontSizes, VisualSizesEnum} from '../../helpers/fontHelpers';
import {makeSizeConstants} from '../../helpers/styleHelpers';

/*
 * Constants.
 */

const avatarSizes = makeSizeConstants(16, 20, 30, 48);
const lineHeights = makeSizeConstants(16, 20, 30, 48);
const avatarFontSizes = makeSizeConstants(fontSizes.tiny, fontSizes.tiny, fontSizes.small, fontSizes.large);
const characterLimits = makeSizeConstants(1, 2);

/*
 * Props.
 */

interface AvatarProps {
  /** The name to use for the avatar. */
  name: string;
  /** If specified, we will render the image instead of initials. */
  imgSrc?: string;
  /** The size of the avatar. Defaults to VisualSizesEnum.LARGE. */
  size?: VisualSizesEnum;
}

/*
 * Style.
 */

interface StyledAvatarWrapperDivProps {
  $color: string;
  $size: VisualSizesEnum;
  $imgSrc?: string;
}

const StyledAvatarWrapperDiv = styled.div<StyledAvatarWrapperDivProps>`
  font-family: ${fonts.system};
  width: ${p => `${avatarSizes[p.$size]}px`};
  height: ${p => `${avatarSizes[p.$size]}px`};
  line-height: ${p => `${lineHeights[p.$size]}px`};
  font-size: ${p => avatarFontSizes[p.$size]};
  background: ${p => p.$color};
  border-radius: 50%;
  text-align: center;
  color: ${greys.white};

  ${p => addImageSrcStyles(p.$imgSrc)};
`;

function addImageSrcStyles(imgSrc?: string) {
  if (!imgSrc)
    return '';
  return css`
    background: transparent;
    background-image: url(${imgSrc});
    background-size: cover;
    background-position: center center;
    image-rendering: -webkit-optimize-contrast;
    image-rendering: crisp-edges;
  `;
}

/*
 * Component.
 */

export const Avatar: FC<AvatarProps> = props => {
  const {name, imgSrc, size = VisualSizesEnum.LARGE} = props;
  const avatarColor = useMemo(() => computeColorFromName(name.trim()), [name]);
  const initials = useMemo(() => computeInitialsFromName(name.trim(), size), [name, size]);

  return <StyledAvatarWrapperDiv $color={avatarColor} $size={size} $imgSrc={imgSrc}>{!imgSrc && initials}</StyledAvatarWrapperDiv>;
};

/*
 * Helpers.
 */

function computeColorFromName(name: string) {
  if (name.length === 0)
    return palette.blue.shade40;
  // Count the list of available colors and remove 1 since we do not allow grey as a default color.
  const availableColorsCount = Object.keys(palette).length - 1;
  const selectedColorIndex = ((name.length * name.charCodeAt(0) * name.charCodeAt(name.length - 1)) % availableColorsCount) + 1;

  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  const selectedColorPalette = Object.keys(palette)[selectedColorIndex] as PaletteColorsEnum;
  return palette[selectedColorPalette].shade40;
}

function computeInitialsFromName(name: string, size: VisualSizesEnum) {
  const initialLimits = characterLimits[size];
  return name.split(' ').map(word => word[0]).slice(0, initialLimits).join('');
}
