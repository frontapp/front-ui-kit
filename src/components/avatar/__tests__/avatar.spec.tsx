import {render} from '@testing-library/react';
import React from 'react';

import {VisualSizesEnum} from '../../../helpers/fontHelpers';
import {Avatar} from '../avatar';

describe('<Avatar /> Spec', () => {
  describe('Snapshots', () => {
    it('should render avatar with name and color should always be the same', () => {
      const view = render(<Avatar name="John Doe" />);
      expect(view.container).toMatchSnapshot();
    });

    it('should render small avatar with a single initial', () => {
      const view = render(<Avatar name="John Doe" size={VisualSizesEnum.SMALL} />);
      expect(view.container).toMatchSnapshot();
    });

    it('should render large avatar', () => {
      const view = render(<Avatar name="John Doe" size={VisualSizesEnum.LARGE} />);
      expect(view.container).toMatchSnapshot();
    });

    it('should render an image avatar', () => {
      const view = render(<Avatar name="John Doe" imgSrc="https://picsum.photos/id/1062/200/200" />);
      expect(view.container).toMatchSnapshot();
    });
  });
});
