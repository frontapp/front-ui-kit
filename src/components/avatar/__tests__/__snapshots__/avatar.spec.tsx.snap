// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Avatar /> Spec Snapshots should render an image avatar 1`] = `
.c0 {
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Helvetica Neue", Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  width: 30px;
  height: 30px;
  line-height: 30px;
  font-size: 13px;
  background: #ca3e99;
  border-radius: 50%;
  text-align: center;
  color: #ffffff;
  background: transparent;
  background-image: url(https://picsum.photos/id/1062/200/200);
  background-size: cover;
  background-position: center center;
  image-rendering: -webkit-optimize-contrast;
  image-rendering: crisp-edges;
}

<div>
  <div
    class="c0"
  />
</div>
`;

exports[`<Avatar /> Spec Snapshots should render avatar with name and color should always be the same 1`] = `
.c0 {
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Helvetica Neue", Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  width: 30px;
  height: 30px;
  line-height: 30px;
  font-size: 13px;
  background: #ca3e99;
  border-radius: 50%;
  text-align: center;
  color: #ffffff;
}

<div>
  <div
    class="c0"
  >
    JD
  </div>
</div>
`;

exports[`<Avatar /> Spec Snapshots should render large avatar 1`] = `
.c0 {
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Helvetica Neue", Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  width: 30px;
  height: 30px;
  line-height: 30px;
  font-size: 13px;
  background: #ca3e99;
  border-radius: 50%;
  text-align: center;
  color: #ffffff;
}

<div>
  <div
    class="c0"
  >
    JD
  </div>
</div>
`;

exports[`<Avatar /> Spec Snapshots should render small avatar with a single initial 1`] = `
.c0 {
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Helvetica Neue", Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  width: 16px;
  height: 16px;
  line-height: 16px;
  font-size: 11px;
  background: #ca3e99;
  border-radius: 50%;
  text-align: center;
  color: #ffffff;
}

<div>
  <div
    class="c0"
  >
    J
  </div>
</div>
`;
